src/key-encoder.ts

```diff
@@ -0,0 +1,183 @@
+    // @ts-ignore
+    const self = this as any
+    self.seq().obj(
+        self.key('algorithm').seq().obj(
+            self.key("id").objid(),
+            self.key("curve").objid()
+        ),
+        self.key('pub').bitstr()
+    )
+})
+
+interface CurveOptions {
+    curveParameters: number[];
+    privatePEMOptions: { label: string };
+    publicPEMOptions: { label: string };
+    curve: EC;
+}
+
+const curves: { [index: string]: CurveOptions } = {
+    secp256k1: {
+        curveParameters: [1, 3, 132, 0, 10],
+        privatePEMOptions: { label: 'EC PRIVATE KEY' },
+        publicPEMOptions: { label: 'PUBLIC KEY' },
+        curve: new EC('secp256k1')
+    }
+}
+
+interface PrivateKey {
+    version: BN;
+    privateKey: Buffer;
+    parameters: number[];
+    publicKey?: {
+        unused: number;
+        data: Buffer;
+    }
+}
+
+export default class KeyEncoder {
+    static ECPrivateKeyASN = ECPrivateKeyASN
+    static SubjectPublicKeyInfoASN = SubjectPublicKeyInfoASN
+
+    algorithmID: number[]
+    options: CurveOptions
+
+    constructor(options: string | CurveOptions) {
+        if (typeof options === 'string') {
+            if (options !== 'secp256k1') {
+                throw new Error('Unknown curve ' + options)
+            }
+            options = curves[options]
+        }
+        this.options = options
+        this.algorithmID = [1, 2, 840, 10045, 2, 1]
+    }
+
+    privateKeyObject(rawPrivateKey: string, rawPublicKey: string) {
+        const privateKeyObject: PrivateKey = {
+            version: new BN(1),
+            privateKey: Buffer.from(rawPrivateKey, 'hex'),
+            parameters: this.options.curveParameters
+        }
+
+        if (rawPublicKey) {
+            privateKeyObject.publicKey = {
+                unused: 0,
+                data: Buffer.from(rawPublicKey, 'hex')
+            }
+        }
+
+        return privateKeyObject
+    }
+
+    publicKeyObject(rawPublicKey: string) {
+        return {
+            algorithm: {
+                id: this.algorithmID,
+                curve: this.options.curveParameters
+            },
+            pub: {
+                unused: 0,
+                data: Buffer.from(rawPublicKey, 'hex')
+            }
+        }
+    }
+
+    encodePrivate(privateKey: string | Buffer, originalFormat: string, destinationFormat: string) {
+        let privateKeyObject: PrivateKey
+
+        /* Parse the incoming private key and convert it to a private key object */
+        if (originalFormat === 'raw') {
+            if (typeof privateKey !== 'string') {
+                throw 'private key must be a string'
+            }
+            let keyPair = this.options.curve.keyFromPrivate(privateKey, 'hex')
+            let rawPublicKey = keyPair.getPublic('hex')
+            privateKeyObject = this.privateKeyObject(privateKey, rawPublicKey)
+        } else if (originalFormat === 'der') {
+            if (typeof privateKey !== 'string') {
```

Comment:
these branches can be eliminated now because of TypeScript's checks, right? So this could be:

```javascript
....
} else if (originalFormat === 'der') {
      if (typeof privateKey === 'string') {
         privateKey = Buffer.from(privateKey, 'hex')
      }
      privateKeyObject = ECPrivateKeyASN.decode(privateKey, 'der')
} else if ...
```

Or does TypeScript not strictly enforce those types here?

Reply:
actually `orginalFormat` looks like it should be a const string enum, or string discriminate union type. will improve

Reply:
Added a type for the valid string literals in these methods. 
This prevents a typescript consumer of this lib from compiling with an invalid `format` option. 

However, regular js consumers can pass in whatever they want. Typescript doesn't inject runtime checks. The compilation output for this code looks like regular js:
```js
    encodePrivate(privateKey, originalFormat, destinationFormat) {
        let privateKeyObject;
        /* Parse the incoming private key and convert it to a private key object */
        if (originalFormat === 'raw') {
            if (typeof privateKey !== 'string') {
                throw 'private key must be a string';
            }
            let keyPair = this.options.curve.keyFromPrivate(privateKey, 'hex');
            let rawPublicKey = keyPair.getPublic('hex');
            privateKeyObject = this.privateKeyObject(privateKey, rawPublicKey);
        }
```

A lot of projects use `typeforce` instead of doing it manually, like here:
https://github.com/bitcoinjs/bitcoinjs-lib/blob/3bee658becce594a57cd853f9933a800986850f1/ts_src/script.ts#L170
```ts
export function fromASM(asm: string): Buffer {
  typeforce(types.String, asm);
  ...
```